fix(core): update `isDevMode` to rely on `ngDevMode`

alan-agius4 opened the pull request:
This commits update `isDevMode` to rely on the `ngDevMode` which in the CLI is set by the bundler.

We also update `@angular/platform-dynamic-browser` and `@angular/compiler` to remove usage of `jitDevMode`, which was the last internal usages of `isDevMode`.

alan-agius4 commented on packages/core/src/util/is_dev_mode.ts:
```diff
@@ -29,7 +29,8 @@ let _runModeLocked: boolean = false;
  */
 export function isDevMode(): boolean {
```
Should this be deprecated or it's still useful as public API?

alan-agius4 commented on packages/core/src/util/is_dev_mode.ts:
```diff
@@ -49,7 +50,7 @@ export function enableProdMode(): void {
   // The below check is there so when ngDevMode is set via terser
   // `global['ngDevMode'] = false;` is also dropped.
-  if (typeof ngDevMode === undefined || !!ngDevMode) {
+  if (typeof ngDevMode === 'undefined' || !!ngDevMode) {
```
Oops

pkozlowski-opensource commented:
We've discussed this on the fwk side in the past and I believe that this would be breaking change. Adding a label accordingly.

alan-agius4 commented:
@pkozlowski-opensource  why would this be a breaking change? 

Calling `enableProdMode` set `ngDevMode` to false. So, why is this different?

pkozlowski-opensource commented:
> The framework doesn't use the isDevMode method.

@alan-agius4 the framework does not, but the existing libraries might (and probably do).

AndrewKushnir commented on packages/core/src/util/is_dev_mode.ts:
```diff
@@ -43,15 +30,9 @@ export function isDevMode(): boolean {
  * @publicApi
  */
 export function enableProdMode(): void {
```
We can probably also update docs for this function to discourage its usage and instead explain how to enable the prod mode via CLI flags (for CLI-based apps).

AndrewKushnir reviewed (approved):
LGTM 👍

It'd be great to add some additional docs for the `isDevMode` and `enableProdMode` functions to explain where the value is coming from and what'd be the best way to change it (via CLI flags vs enabling at runtime).

alan-agius4 committed:
fix(core): update `isDevMode` to rely on `ngDevMode`

This commits update `isDevMode` to rely on the `ngDevMode` which in the CLI is set by the bundler.

We also update `@angular/platform-dynamic-browser` and `@angular/compiler` to remove usage of `jitDevMode`, with this change we remove all internal usages of `isDevMode`.

alxhub commented:
This PR was merged into the repository by commit 85330f3fd9ac6381ce3aa18479ed8195d2ac215e.

Platonn commented on packages/core/src/util/is_dev_mode.ts:
```diff
@@ -9,27 +9,16 @@
  */
 export function isDevMode(): boolean {
-  _runModeLocked = true;
-  return _devMode;
+  return typeof ngDevMode === 'undefined' || !!ngDevMode;
 }
```
Hi @alan-agius4. I have a question, so to better understand:
Now `isDevMode()` explicitly depends on `ngDevMode`. Does it by any chance help to tree-shake code in custom applications, that is conditionally wrapped with `isDevMode`()? e.g.
```ts
if(isDevMode()){ 
   console.error('Some very long string message in custom app....');
}
```

If not, would it help to expose `ngDevMode` in Angular public API, so it could be used it in custom applications?
```ts
if(ngDevMode){ 
   console.error('Some very long string message in custom app....');
}
```

eneajaho replied:
Just tested this isDevMode(). It doesn't help tree-shake code. 


<img width="327" alt="Screenshot 2022-09-29 at 12 29 55 PM" src="https://user-images.githubusercontent.com/25394362/193008388-aa0082e8-1607-4ab4-99a7-54ce262edd6a.png">

<img width="382" alt="Screenshot 2022-09-29 at 12 30 09 PM" src="https://user-images.githubusercontent.com/25394362/193008446-43669d5c-70d7-4891-bad3-c406e10f667c.png">

alan-agius4 replied:
Indeed this does not aid tree-shaking.

`ngDevMode` is still considered as a private API, but we do have a project to make this available as a public API.
